### PR TITLE
Fixes #11586 - Hammer CSV output saved as a file

### PR DESCRIPTION
**Pull request by ofedoren**

### CODE REVIEW

**Comment by tstrachota** [comment]
[test]

**Comment by mbacovsky** [comment]
[test]

**Comment by mbacovsky** [comment]
@ofedoren thanks for the PR! I have some questions - 
 - the original BZ and probably also the redmine issue talks about hammer csv plugin and seems to be resolved already. Do we have demand for output to file in hammer in general ( @tstrachota )? While it brings some complexity into adapters I'd like to have confirmation it is needed.
 - I had a problem with making the feature to work:
```
>  hammer --csv --output-file /tmp/test.csv -s https://<fqdn> architecture list
Could not load the API description from the server: 
Failed to open TCP connection to foreman.my.lan:80 (getaddrinfo: Name or service not known)
  - is the server down?
  - was 'foreman-rake apipie:cache' run on the server when using apipie cache? (typical production settings)
```
While without the `--output-file ...` it worked. Did you meet this @ofedoren? It prevented me from further testing so far. Once this is solved I'd like to check how it behaves in less expected cases (create, pagination, interactive mode)
 - If we decide to include this in, I'd like to have the adapter#output stream covered by some test. Individual adapters might be worth some testing too.

 I'll keep investigating the failures.

**Comment by mbacovsky** [comment]
I found out if I swap `--output-file` and `-s` options it works well. I'll have to check what is going on there.

**Comment by mbacovsky** [comment]
So the problem was the `--output-file` option needs to be defined also in `bin/hammer` otherwise the pre-parser fails on unknown option and will not parse the other args.

**Comment by ofedoren** [comment]
@mbacovsky, updated. Added some tests.

**Comment by tstrachota** [comment]
@mbacovsky ping. What's the status here?

File: lib/hammer_cli/main.rb [hunk 1]
```diff
@@ -44,6 +44,13 @@ class MainCommand < AbstractCommand
     option ["--output"], "ADAPTER", _("Set output format. One of [%s]") %
       HammerCLI::Output::Output.adapters.keys.join(', '),
       :context_target => :adapter
+    option ["--output-file"], "OUTPUT_FILE", _("path to custom output file") do |filename|
+      context[:output_file] = File.new(filename, "w")
+      unless File.writable? context[:output_file]
```
[the remark this answers is not in the document]
Comment (mbacovsky) [reply lib/hammer_cli/main.rb]: Also when the file is not writable e.g. due to permissions or missing dir in path it raises an exception during `File.new(filename, "w")`. What conditions should the `writable?` test cover?

**Comment by mbacovsky** [comment]
@ofedoren , could you rebase, please?

**Comment by ofedoren** [comment]
@mbacovsky , done. But I don't think that the handler's output actually tells user to check the file's permissions.

```
> hammer --csv --output-file /tmp/test.csv -s https://<fqdn> architecture list
Error: Option '--output-file': ArgumentError.

See: 'hammer --help'.
```

**Comment by tstrachota** [comment]
@mbacovsky ping. Can you re-review, please?

**Comment by mbacovsky** [comment]
[test]

**Comment by mbacovsky** [comment]
@ofedoren, thanks for the update. Regarding the error handling this is what I'm getting and it is what I expected:
```
$ hammer --csv --output-file /bin/out.csv architecture list 
Error: Option '--output-file': Permission denied.

See: 'hammer --help'.
```
One question though - why do you strip the filename from the message?

**Comment by ofedoren** [comment]
@mbacovsky updated. I decided to leave the filename in error message.

**Comment by mbacovsky** [comment]
[test]

**Comment by mbacovsky** [comment]
@ofedoren, thanks for the update. All look well now. Merging

**Pull request merged**